packages/bull-vault/src/FlashBull.sol

```diff
@@ -1,40 +1,207 @@
+    /// @dev ETH:wSqueeth Uniswap pool
+    address private immutable ethWSqueethPool;
+    /// @dev ETH:USDC Uniswap pool
+    address private immutable ethUSDCPool;
+    /// @dev Crab V2 address
+    address private immutable crab;
+    /// @dev Uniswap oracle address
+    address public immutable oracle;
+
+    address public bullStrategy;
+
+    struct FlashDepositCrabData {
+        uint256 ethToDepositInCrab;
+    }
+
+    struct FlashDepositCollateralData {
+        uint256 crabToDeposit;
+        uint256 ethToLend;
+    }
+
+    constructor(address _bull, address _factory, address _oracle) UniBull(_factory) {
+        bullStrategy = _bull;
+        oracle = _oracle;
+        crab = IBullStrategy(_bull).crab();
+        wPowerPerp = IController(IBullStrategy(bullStrategy).powerTokenController()).wPowerPerp();
+        weth = IController(IBullStrategy(bullStrategy).powerTokenController()).weth();
+        usdc = IController(IBullStrategy(bullStrategy).powerTokenController()).quoteCurrency();
+        ethWSqueethPool = IController(IBullStrategy(bullStrategy).powerTokenController()).wPowerPerpPool();
+        ethUSDCPool = IController(IBullStrategy(bullStrategy).powerTokenController()).ethQuoteCurrencyPool();
+    }
+
+    /**
+     * @notice receive function to allow ETH transfer to this contract
+     */
+    receive() external payable {
+        require(msg.sender == weth, "Message sender should be WETH");
+    }
+
+    /**
+     * @notice flash deposit into strategy, providing ETH, selling wSqueeth and dollars, and receiving strategy tokens
+     * @dev this function will execute a flash swap where it receives ETH, deposits, mints, and collateralizes the loan using flash swap proceeds and msg.value, and then repays the flash swap with wSqueeth and USDC
+     * @param _ethToCrab ETH that will be deposited into the crab strategy
+     * @param _poolFee Uniswap pool fee
+     */
+    function flashDeposit(uint256 _ethToCrab, uint24 _poolFee) external payable {
+        uint256 squeethEthPrice = _getTwap(ethWSqueethPool, wPowerPerp, weth, TWAP, false);
+        uint256 crabAmount;
+        uint256 usdcToBorrow;
+        uint256 ethToLend;
+        uint256 wSqueethToMint;
+        {
+            (uint256 ethInCrab, uint256 squeethInCrab) = IBullStrategy(bullStrategy).getCrabVaultDetails();
+            uint256 ethUsdPrice = _getTwap(ethUSDCPool, weth, usdc, TWAP, false);
+            uint256 crabUsdPrice = (
+                ethInCrab.wmul(ethUsdPrice).sub(squeethInCrab.wmul(squeethEthPrice).wmul(ethUsdPrice))
+            ).wdiv(IERC20(crab).totalSupply());
+            uint256 ethFee;
+            (wSqueethToMint, ethFee) = _calcWsqueethToMintAndFee(_ethToCrab, squeethInCrab, ethInCrab, squeethEthPrice);
+            crabAmount = _calcSharesToMint(_ethToCrab.sub(ethFee), ethInCrab, IERC20(crab).totalSupply());
+
+            uint256 share;
+            if (IERC20(bullStrategy).totalSupply() == 0) {
+                share = ONE;
```

Comment:
dis isnt same as deposit() see how its done. there is also a security risk in the way we do it!

Reply:
maybe rename to bullOwnershipShare or something - to me it seems too close to bull shares.  in the deposit function we also have shares that are minted, which aren't the same